lib/Frontend/DiagnosticVerifier.cpp

```diff
@@ -786,42 +771,49 @@ void DiagnosticVerifier::autoApplyFixes(unsigned BufferID,
 /// Every time a diagnostic is generated in -verify mode, this function is
 /// called with the diagnostic.  We just buffer them up until the end of the
 /// file.
-static void VerifyModeDiagnosticHook(const llvm::SMDiagnostic &Diag,
-                                     void *Context) {
-  ((DiagnosticVerifier*)Context)->addDiagnostic(Diag);
-}
+void DiagnosticVerifier::handleDiagnostic(SourceManager &SM,
+                                          const DiagnosticInfo &Info) {
+  CapturedDiagnosticInfo capturedInfo;
+  capturedInfo.Classification = Info.Kind;
+  capturedInfo.Loc = Info.Loc;
+  if (Info.Loc.isValid()) {
+    auto lineAndColumn = SM.getLineAndColumn(Info.Loc);
+    capturedInfo.Line = lineAndColumn.first;
+    capturedInfo.Column = lineAndColumn.second;
+    capturedInfo.FileName = SM.getDisplayNameForLoc(Info.Loc);
+  }
 
+  for (auto fixIt : Info.FixIts)
+    capturedInfo.FixIts.push_back(fixIt);
 
-void swift::enableDiagnosticVerifier(SourceManager &SM) {
-  SM.getLLVMSourceMgr().setDiagHandler(VerifyModeDiagnosticHook,
-                                       new DiagnosticVerifier(SM));
+  {
+    llvm::raw_svector_ostream Out(capturedInfo.Message);
+    DiagnosticEngine::formatDiagnosticText(Out, Info.FormatString,
+                                           Info.FormatArgs);
+  }
+
+  CapturedDiagnostics.push_back(capturedInfo);
```

Comment:
Now that you've got this all nicely factored out, could you please compute the members of `CapturedDiagnosticInfo` separately and emplace them? This'll let you reduce the scope of the mutations being performed on the temporary, const-qualify some stuff, and make it clear which parts are contributing to the captured diagnostic and which parts are contributing to the formatting.